### ISSUE TITLE
Bug: Fix Data Races

### ISSUE THREAD

**Issue by peake100** [peake100/rogerRabbit-go#26]
CI pipeline is periodically failing on the event relay tests for events like NotifyReturn and NotifyConfirm tests. There is likely a race condition either in the test code or lib code that we need to track down.

These tests always pass on my laptop. Normally this implies a race condition is the issue, as the CI machines are much beefier.

**Comment by peake100** [comment]
After further investigation running go test with the -race flag, we need to track down and squash some race conditions.

**Comment by peake100** [comment]
Upon further inspection, it appears that most of the data races are caused by logging to stdout as the test is closing and writing its own output.

This is going to be a difficult issue to solve, our test connections and channels are still running in other routines logging away in the background when a suite sub-test method completes. We could disable logging entirely during tests, but end up loosing a lot of rich data.

I'm not sure what the best approach to this is. We may need to restructure the tests so that all connections close between tests. It would be nice to enable the -race flag for CI/CD

**Comment by peake100** [comment]
Turning off all logs, there seems to be one main data race, though I am having trouble interpreting it.

Full Printout:

```
==================
WARNING: DATA RACE
Read at 0x00c0001606b0 by goroutine 86:
  github.com/peake100/rogerRabbit-go/amqp.(*transportManager).reconnectListenForClose()
      rogerRabbit-go/amqp/transportManagerReconnect.go:57 +0x9a

Previous write at 0x00c0001606b0 by goroutine 88:
  github.com/peake100/rogerRabbit-go/amqp.transportHandlersBuilder.createClose.func1.1()
      rogerRabbit-go/amqp/transportManagerHandlersBuilder.go:203 +0x16b
  github.com/peake100/rogerRabbit-go/amqp.transportHandlersBuilder.createClose.func1()
      rogerRabbit-go/amqp/transportManagerHandlersBuilder.go:212 +0x175
  github.com/peake100/rogerRabbit-go/amqp.transportManager.Close()
      rogerRabbit-go/amqp/transportManager.go:467 +0x76
  github.com/peake100/rogerRabbit-go/amqptest.(*AmqpSuite).ReplaceChannels()
      rogerRabbit-go/amqptest/testSuite.go:202 +0x5c4
  github.com/peake100/rogerRabbit-go/amqptest.(*AmqpSuite).ReplaceChannels-fm()
      rogerRabbit-go/amqptest/testSuite.go:200 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/go/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/go/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/go/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1198 +0x22a

Goroutine 86 (running) created at:
  github.com/peake100/rogerRabbit-go/amqp.(*transportManager).reconnect()
      rogerRabbit-go/amqp/transportManagerReconnect.go:90 +0x216
  github.com/peake100/rogerRabbit-go/amqp.(*transportManager).reconnectListenForClose()
      rogerRabbit-go/amqp/transportManagerReconnect.go:66 +0x192

Goroutine 88 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1239 +0x5d7
  github.com/stretchr/testify/suite.runTests()
      /Users/williampeake/go/pkg/mod/github.com/stretchr/testify@v1.7.0/suite/suite.go:203 +0xf7
  github.com/stretchr/testify/suite.Run()
      /Users/williampeake/go/pkg/mod/github.com/stretchr/testify@v1.7.0/suite/suite.go:176 +0x944
  github.com/peake100/rogerRabbit-go/amqp_test.TestChannelMethods()
      rogerRabbit-go/amqp/channel_test.go:1948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
```

The conflict seems to be between here:

rogerRabbit-go/amqp/transportManagerReconnect.go:57
```go
// reconnectListenForClose listens for a close event from the underlying livesOnce, and
// starts the reconnection process.
func (manager *transportManager) reconnectListenForClose(
	closeChan <-chan *streadway.Error,
) {
	// Wait for the current connection to close
	disconnectEvent := <-closeChan
	// Send a disconnect event to all interested subscribers.
	manager.sendDisconnectNotifications(disconnectEvent) // <--- (read)
...
```
and here:

rogerRabbit-go/amqp/transportManagerHandlersBuilder.go:203
```go
// createClose creates the base handler for transportManager.Close.
func (builder transportHandlersBuilder) createClose() amqpmiddleware.HandlerClose {
	manager := builder.manager

	handler := func(ctx context.Context, args amqpmiddleware.ArgsClose) error {

		// If the context has already been cancelled, we can exit.
		if manager.ctx.Err() != nil {
			return streadway.ErrClosed
		}

		manager.cancelCtxCloseTransport()

		func() {
			// We need to acquire and release the lock here to avoid race conditions.
			manager.notificationSubscriberLock.Lock()
			defer manager.notificationSubscriberLock.Unlock()

			// Close all disconnect and connect subscribers, then clear them. We don't
			// need to grab the lock for this since the cancelled context will keep any new
			// subscribers from being added.
			for _, subscriber := range manager.notificationSubscribersDial {
				close(subscriber)
			}
			// We clear these to avoid sending on a nil channel.
			manager.notificationSubscribersDial = nil // <- previousWrite
```

I'm not quite sure what is is in conflict here, as these resources do not seem to collide. Maybe it has something to do with what get's captured by the close handler? Very odd.

**Comment by peake100** [comment]
manager.notificationSubscribersDial is used by manager.sendDisconnectNotifications, but both routines must hold an exclusive lock to access the data, so I am not sure what is being affected here.

**Comment by peake100** [comment]
Ah, I've figured it out.

Even though the methods are both holding a lock, the one sending the notifications was a pass-by-value of the manager, so if we passed the manager by value, tried to acquire the lock, an during that process the shutdown method already had the lock held and nilled the subscriber list, then the list would still be populated in the first routine and would cause it to try sending on closed channels after acquiring the lock.

**Comment by peake100** [comment]
We must fix it by making sendDisconnectNotifications a pass-by-reference method receiver.

**Comment by peake100** [comment]
Closed by #33
